jetty-util/src/main/java/org/eclipse/jetty/util/Index.java

```diff
@@ -0,0 +1,353 @@
+     *
+     * @return true if the index does not contain any entry.
+     */
+    boolean isEmpty();
+
+    /**
+     * Get the number of entries in the index.
+     *
+     * @return the index' entries count.
+     */
+    int size();
+
+    /**
+     * Get a {@link Set} of the keys contained in this index.
+     *
+     * @return a {@link Set} of the keys contained in this index.
+     */
+    Set<String> keySet();
+
+    /**
+     * A mutable String lookup data structure.
+     * Implementations are not thread-safe.
+     * @param <V> the entry type
+     */
+    interface Mutable<V> extends Index<V>
+    {
+        /**
+         * Put an entry into the index.
+         *
+         * @param s The key for the entry
+         * @param v The value of the entry
+         * @return True if the index had capacity to add the field.
+         */
+        boolean put(String s, V v);
+
+        /**
+         * Put a value as both a key and a value.
+         *
+         * @param v The value and key
+         * @return True if the Trie had capacity to add the field.
+         */
+        boolean put(V v);
+
+        /**
+         * Remove an entry from the index.
+         *
+         * @param s The key for the entry
+         * @return The removed value of the entry
+         */
+        V remove(String s);
+
+        /**
+         * Remove all entries from the index.
+         */
+        void clear();
+
+        /**
+         * Builder of {@link Index.Mutable} instances. Such builder cannot be
+         * directly created, it is instead returned by calling {@link Index.Builder#mutable()}.
+         * @param <V> the entry type
+         */
+        class Builder<V> extends Index.Builder<V>
+        {
+            private int maxCapacity = -1;
+
+            Builder(boolean caseSensitive, Map<String, V> contents)
+            {
+                super(caseSensitive, contents);
+            }
+
+            /**
+             * Configure a maximum capacity for the mutable index.
+             * A negative value means there is no capacity limit and
+             * the index can grow without limits.
+             * The default value is -1.
+             * @param capacity the maximum capacity of the index.
+             * @return this
+             */
+            public Builder<V> maxCapacity(int capacity)
+            {
+                this.maxCapacity = capacity;
+                return this;
+            }
+
+            /**
+             * Build a {@link Mutable} instance.
+             * @return a {@link Mutable} instance.
+             */
+            public Mutable<V> build()
+            {
+                if (contents != null && maxCapacity == 0)
+                    throw new IllegalStateException("Cannot create a mutable index with maxCapacity=0 and some contents");
+
+                Mutable<V> result;
+                if (maxCapacity > 0)
+                    result = new ArrayTernaryTrie<>(!caseSensitive, maxCapacity);
```

Comment:
Is there a max capacity to ArrayTernaryTrie?  If so we might need to pick which implementation

Reply:
I think ArrayTernaryTrie also has character indexes... so it too will have a max size.   Thus we need to check for the max size and use a TreeTrie

Reply:
oops you deleted TreeTrie... I think there is still a case to use it.

Reply:
There is, and I've added a simple check that throws an exception when we go over it. Simplistic, doesn't solve any problem, but at least the problem reported in #5291 would make the code throw a exception with a human-readable error.

As discussed, properly solving that would require a new case-sensitive trie implementation that supports larger sizes, so that's for another PR.